formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt

```diff
@@ -249,58 +456,92 @@ internal open class CborReader(private val cbor: Cbor, protected val decoder: Cb
             val knownIndex: Int
             while (true) {
                 if (isDone()) return CompositeDecoder.DECODE_DONE
-                val elemName = decoder.nextString()
+                val (elemName, tags) = decodeElementNameWithTagsLenient(descriptor)
                 readProperties++
 
-                val index = descriptor.getElementIndex(elemName)
+                val index = elemName?.let { descriptor.getElementIndex(it) } ?: CompositeDecoder.UNKNOWN_NAME
                 if (index == CompositeDecoder.UNKNOWN_NAME) {
-                    decoder.skipElement()
+                    decoder.skipElement(tags)
                 } else {
+                    if (cbor.verifyKeyTags) {
+                        descriptor.getKeyTags(index)?.let { keyTags ->
+                            if (!(keyTags contentEquals tags)) throw CborDecodingException("CBOR tags $tags do not match declared tags $keyTags")
```

Comment:
General observation (for all `throw` expressions): It is would be much easier to debug errors if they also included locations. Here we don't have a json path analog, but appending `descriptor.toString()` will already make the job easier significantly.

Reply:
done where we have access to the descriptor (which we do not always have)